plonky2/Cargo.toml

```diff
@@ -11,45 +11,46 @@ edition = "2021"
 default-run = "generate_constants"
 
 [features]
-default = ["parallel", "rand", "rand_chacha", "timing", "gate_testing"]
-parallel = ["maybe_rayon/parallel"]
-rand = ["dep:rand", "plonky2_field/rand"]
-gate_testing = ["rand"]
-rand_chacha = ["dep:rand_chacha"]
-timing = []
+default = ["gate_testing", "parallel", "rand_chacha", "std", "timing"]
+gate_testing = []
+parallel = ["hashbrown/rayon", "maybe_rayon/parallel"]
+std = ["anyhow/std", "rand/std"]
+timing = ["std"]
 
 [dependencies]
-plonky2_field = { path = "../field" }
-plonky2_util = { path = "../util" }
-log = "0.4.14"
-itertools = "0.10.0"
-num = { version = "0.4", features = [ "rand" ] }
-rand = { version = "0.8.4", optional = true }
-rand_chacha = { version = "0.3.1", optional = true }
-maybe_rayon = { path = "../maybe_rayon" }
-unroll = "0.1.5"
-anyhow = "1.0.40"
-serde = { version = "1.0", features = ["derive"] }
-serde_cbor = "0.11.1"
-keccak-hash = "0.8.0"
-static_assertions = "1.1.0"
+anyhow = { version = "1.0.40", default-features = false }
+derivative = { version = "2.2.0", default-features = false, features = ["use_core"] }
+hashbrown = { version = "0.12.3", default-features = false, features = ["ahash", "serde"] }
```

Comment:
The lack of DOS resistance seems a bit concerning, but I think it should be okay... I can't think of a case where we have a potentially-huge `HashMap` with untrusted input. The largest `HashMap`s should be in `CircuitBuilder`, thankfully we no longer use `HashMap` in witness generation.

Reply:
Why don't we enable `ahash-compile-time-rng` just in case? So at least users who compile their own binaries will be protected.

I wonder why it's not the default, maybe some users expect hashbrown's `HashMap` to be deterministic? But we certainly don't since std `HashMap` isn't.

Reply:
Yeah I chose AHash because of its performance given that I don't think that the DOS resistance is that much of a risk in the plonky2 use case. Added `compile-time-rng` though.

Reply:
I don't think adding our own `ahash` dependency with `compile-time-rng` affects `hashbrown`; `cargo tree` is showing separate copies of `ahash`. I thought enabling `ahash-compile-time-rng` in hashbrown was the right approach, but it seems like that feature was removed in https://github.com/rust-lang/hashbrown/pull/357.

It seems the suggestion now is to do something like `hashbrown::HashMap::with_hasher(ahash::RandomState::new())` when we build a `HashMap` that needs HashDoS resistance. Since we don't have a clear need for HashDoS resistance currently, I think it's okay to just go with the default deterministic behavior for now.

Reply:
> I don't think adding our own ahash dependency with compile-time-rng affects hashbrown; cargo tree is showing separate copies of ahash. I thought enabling ahash-compile-time-rng in hashbrown was the right approach, but it seems like that feature was removed in https://github.com/rust-lang/hashbrown/pull/357.

Ahh, sorry I forgot to match the version from the `hashbrown` dependency. Because features are additive as long as you match the same version you can include more features in the dependency (even if it's a sub-dependency). The drawback of this approach is that we have to keep the version synchronized when upgrading the `hashbrown` dependency.

Reply:
Ah ok, seems like a pretty good solution then